examples/dreamcast/basic/threading/reentrant_mutex/reentrant_mutex.c

```diff
@@ -0,0 +1,179 @@
+            fprintf(stderr, "thread %s: rmutex->count was %u when it MUST be zero!\n",
+                    thd_get_label(thd_current), rmutex->count);
+            exit(EXIT_FAILURE);
+        }
+        rmutex->count++;
+    }
+}
+
+/* Unlocks our reentrant_mutex structure. */
+static void reentrant_mutex_unlock(reentrant_mutex_t *rmutex) {
+    /* We better currently be the owning thread if we're attempting to unlock it! */
+    if(atomic_load(&rmutex->owner) == (_Atomic kthread_t *)thd_current) {
+        /* Decrement the lock counter. */
+        if(atomic_fetch_sub(&rmutex->count, 1) == 1) {
+            /* Clear the owning thread and release the mutex if counter hits 0. */
+            atomic_store(&rmutex->owner, NULL);
+            mutex_unlock(&rmutex->mutex);
+        }
+
+    } else {
+        fprintf(stderr, "Error: Thread does not own the mutex\n");
+    }
+}
+
+/* Use the provided probability to determine whether the current thread
+   should pass/yield execution to another thread. */
+static void maybe_pass(void) {
+    unsigned value;
+
+    getentropy(&value, sizeof(value));
+    
+    value %= 100;
+    if(value < THREAD_PASS_CHANCE)
+        thd_pass();
+}
+
+/* Mutex controlling access to "shared_variable." */
+static reentrant_mutex_t rmutex;
+/* Variable all threads are fighting over modifying. */
+static int shared_variable = 0;
+ 
+/* Exec function for each thread. Attempts to acquire the reentrant_mutex
+   multiple times, potentially yielding in between each acquisition,
+   and incrementing the shared variable once while the thread owns the lock. */
+static void *thread_func(void *arg) {
+    printf("Hello from thread %s!\n",
+           thd_get_label(thd_current));
+ 
+    maybe_pass();
+    reentrant_mutex_lock(&rmutex);    // lock count = 1
+ 
+    maybe_pass();
+    reentrant_mutex_lock(&rmutex);    // lock count = 2
+ 
+    shared_variable++;
+ 
+    maybe_pass();
+    reentrant_mutex_unlock(&rmutex);  // lock count = 1
+ 
+    maybe_pass();
+    reentrant_mutex_lock(&rmutex);    // lock count = 2
+ 
+    maybe_pass();
+    reentrant_mutex_unlock(&rmutex);  // lock count = 1
+ 
+    maybe_pass();
+    reentrant_mutex_unlock(&rmutex);  // unlocked
+ 
+    return NULL;
+}
+
+int main(int argc, const char* argv[]) {
+    kthread_t *threads[THREAD_COUNT];
+    
+    /* Initialize our mutex */
+    reentrant_mutex_init(&rmutex);
+ 
+    /* Spawn a bunch of threads, potentially yielding the main thread after
+       each one gets spawned. */
+    for(size_t i = 0; i < THREAD_COUNT; ++i) {
+        threads[i] = thd_create(false, thread_func, NULL);
+        char thd_label[16];
+        snprintf(thd_label, sizeof(thd_label), "%u", i);
+        thd_set_label(threads[i], thd_label);
+ 
+        maybe_pass();
+    }
+ 
+    /* Wait for each thread to complete. */
+    for(size_t i = 0; i < THREAD_COUNT; ++i)
+        thd_join(threads[i], NULL);
+
+    /* Clean up our mutex. */
+    reentrant_mutex_uninit(&rmutex);
```

Comment:
Should there be a check here also that count == 0 and owner == NULL to test that? Or since that's behavior of the upper layer (and not the kos mutex) does it not really matter?

Reply:
Same might go for error checking the mutex_lock/unlock or validating owner == mutex.owner in the lock/unlock. Not too familiar with the exact issues this is intended to validate. I'm guessing the circumstances that would cause those kinds of errors are fairly far from what this is meant to catch.

Reply:
AH! You're absolutely right! There is totally more validation I can do for these scenarios... Just did it plus cleaned up a bit more. Good call.